tests/integration/test_headers_and_subtotals.py

```diff
@@ -3810,3 +3813,457 @@ def it_provide_residual_test_stats_including_hs(self):
         np.testing.assert_array_equal(slice_.zscores, slice_.residual_test_stats[1])
 
         assert slice_.residual_test_stats.shape == (2, 6, 7)
+
+
+class DescribeIntegrated_SubtotalDifferences(object):
+    """TDD driver(s) for Subtotal Difference insertions."""
+
+    def test_cat_subdiff_count(self, row_subdiff_insertions):
+        strand = Cube(CR.CAT_SUBDIFF, transforms=row_subdiff_insertions).partitions[0]
+        counts = strand.counts
+        # --- Subdiff in position 0 equals id=1 - (id=4 + id=5) ---
+        np.testing.assert_almost_equal(counts[0], counts[1] - (counts[4] + counts[5]))
```

Comment:
Just over the past couple weeks we've moved away from `np.testing.*` matchers like `.assert_almost_equal()`. Slobodan initiated the move, but I agree it's more readable and generally more in line with our overall use of `pytest`.

There are two variants that I've encountered, depending on whether the expectation is integer or float in base type.

`.counts` is `float` I believe, so this one would be:

```
assert counts[0] == pytest.approx(counts[1] - (counts[4] + counts[5]))
```

Just as I'm looking at it more closely here, I'd be inclined to make it:

```
assert counts[0] == pytest.approx(142.7) (or whatever the right answer is)
```

rather than duplicating the logic in the test.

The other case, for integer results (which have a solid equivalence comparison) is:

```
assert bases.tolist() == [1, 2, 3, 4, 5]
```

Using these two, you can cover almost all cases, however when NaNs are involved, you need:

```
assert zscores == pytest.approx([1.1, 2.2, np.nan, 4.4], nan_ok=True)
```

Reply:
For my curiosity, can you describe why it's more readable? Just the fact that it uses the `==` infix operator? To me it seems a shame to lose all the niceties of np's wisdom of numeric comparisons.

Reply:
Well, readability is at least partially subjective, but from my own perspective, I'd say that it's the fact that all assertions declare themselves right up front with an `assert ` keyword. A test naturally has three blocks, the fixturing, the call, and the assertions, and having them stand out like that makes them easier to detect while scanning (which is (or should be) much of what we call "reading" code).

@slobodan-ilic might have a different view he could share with us.